synapse/util/caches/sync_response_cache.py

```diff
@@ -0,0 +1,101 @@
+import logging
+from typing import Any, Callable, Dict
+
+from twisted.internet import defer
+
+from synapse.util import Clock
+from synapse.util.async_helpers import ObservableDeferred
+from synapse.util.caches.response_cache import ResponseCache, T
+
+logger = logging.getLogger(__name__)
+
+
+# A special class for /sync responses, to conditionally cache these.
+class SyncResponseCache(ResponseCache[T]):
+    def __init__(self, clock: Clock, name: str, timeout_ms: float = 0):
+        super().__init__(clock, name, timeout_ms)
+
+        self.conditionals = {}  # type: Dict[T, Callable[[Any], bool]]
+
+    def run_conditional(self, key: T, result: Any) -> bool:
+        """Runs a conditional set on key T, defaults to True"""
+        cond = self.conditionals.get(key, None)
+        if cond is None:
+            return True
+        else:
+            try:
+                # Below type annotation is needed for mypy to shush about some statements being unreachable,
+                # we essentially have to not trust other functions to be able to correctly recover from any fallout
+                # (and log it)
+                res = cond(result)  # type: Any
+            except Exception:
+                logger.exception(
+                    "[%s]: Executing conditional %r on %s raised an exception.",
+                    self._name,
+                    cond,
+                    key,
+                )
+                # Evict cache out of caution.
+                return False
+            else:
+                if not isinstance(res, bool):
+                    logger.warning(
+                        "[%s]: Conditional %r returned non-bool value %r (for key %r)",
+                        self._name,
+                        cond,
+                        res,
+                        key,
+                    )
+                    # Return concrete boolean value based on falsy or truthiness.
+                    # If this raises, then so be it, then this value wasn't ever supposed to be true" or "false"
+                    # anyways, then have it be a scream test.
```

Comment:
I could be more correct by wrapping a `try`-`catch` around this, but i think that'd be a bit too over-cautious, i dont know what situations could cause `bool()` to raise, but coupled with the limited use, i think this is fine.

Still added a comment, though.

Reply:
(Yes, i'd like it to be *this* "correct")